### PR TITLE
Allow configuring multiple SPs in IAS. Configure SP for SKR dex

### PR DESCRIPTION
**Pull request by ebensom**
**Description**

Changes proposed in this pull request:

- Add ability to configure multiple SPs in IAS based on a predefined SP input. 
- Configure SP for SKR dex aloing with grafana

### CODE REVIEW

File: components/kyma-environment-broker/internal/process/provisioning/ias_type.go [hunk 1]
```diff
@@ -35,16 +35,21 @@ func (s *IASType) ConfigureType(operation internal.ProvisioningOperation, runtim
...
 
-	spb := s.bundleBuilder.NewBundle(operation.InstanceID)
-	err := spb.FetchServiceProviderData()
-	if err != nil {
-		return s.handleError(operation, err, log, "fetching ServiceProvider data failed")
-	}
+	for spID := range ias.ServiceProviderInputs {
+		spb, err := s.bundleBuilder.NewBundle(operation.InstanceID, spID)
+		if err != nil {
+			return s.handleError(operation, err, log, "failed to create ServiceProvider Bundle")
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/process/provisioning/ias_type.go:41]: Like in `ias_registration` you can directly return `operation, 0, nil`, and log about the error.  Why not return an error? It is a step after success provisioning SKR and we don't want to block the whole operation, log with error will be informed about failure. (in future notification about this kind of failure will be reported)

File: components/kyma-environment-broker/internal/ias/bundle.go [hunk 2]
```diff
@@ -124,22 +128,51 @@ func (b *ServiceProviderBundle) DeleteServiceProvider() error {
...
+}
+
 // ConfigureServiceProviderType sets SSO type, name and URLs based on provided URL for ServiceProvider
 func (b *ServiceProviderBundle) ConfigureServiceProviderType(consolePath string) error {
 	u, err := url.ParseRequestURI(consolePath)
 	if err != nil {
 		return errors.Wrap(err, "while parsing path for IAS Type")
 	}
-	path := strings.Replace(u.Host, "console.", "grafana.", 1)
+	b.serviceProviderDNS = strings.Replace(u.Host, "console.", fmt.Sprintf("%s.", b.serviceProviderParams.Domain), 1)
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/ias/bundle.go:164]: why does it struct parameter if it is used only in this method? It is not better to pass the DNS(name) as an argument?

File: components/kyma-environment-broker/internal/ias/bundle_test.go [hunk 3]
```diff
@@ -50,7 +50,7 @@ func TestServiceProviderBundle_ConfigureServiceProviderType(t *testing.T) {
 
 	// then
 	assert.NoError(t, err)
-	provider, err := client.GetServiceProvider(FakeProviderID)
+	provider, err := client.GetServiceProvider(FakeGrafanaID)
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/ias/bundle_test.go:53]: missing test for `ConfigureServiceProviderType` method for `saml2` type

File: components/connectivity-adapter/go.sum [hunk 4]
```diff
@@ -103,8 +103,10 @@ github.com/kr/pty v1.1.1/go.mod h1:pFQYn66WHrOpPYNljwOMqo10TkYh1fy3cYio2l3bCsQ=
 github.com/kr/text v0.1.0 h1:45sCR5RtlFHMR4UwH9sdQ5TC8v0qDQCHnXt+kaKSTVE=
```
[the remark this answers is not in the document]
Comment (ebensom) [reply components/connectivity-adapter/go.sum]: Removed.

File: components/kyma-environment-broker/internal/ias/input.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		redirectPath:  "/callback",
+		allowedGroups: []string{"runtimeOperator", "runtimeAdmin"},
+		overrides:     nil,
+	},
+	"grafana": {
+		Domain:        "grafana",
+		ssoType:       "openIdConnect",
+		redirectPath:  "/login/generic_oauth",
+		allowedGroups: []string{"skr-monitoring-admin", "skr-monitoring-viewer"},
+		overrides: func(b *ServiceProviderBundle) (string, []*gqlschema.ConfigEntryInput) {
```
[the remark this answers is not in the document]
Comment (ebensom) [reply components/kyma-environment-broker/internal/ias/input.go]: 👍

File: components/kyma-environment-broker/internal/ias/bundle.go [hunk 6]
```diff
@@ -124,22 +128,51 @@ func (b *ServiceProviderBundle) DeleteServiceProvider() error {
...
+	}
+
+	return b.client.SetOIDCConfiguration(b.serviceProvider.ID, iasType)
+}
+
+func (b *ServiceProviderBundle) configureServiceProviderSAMLType(redirectURI string) error {
+	iasType := SAMLType{
+		ServiceProviderName: b.serviceProviderDNS,
+		ACSEndpoints: []ACSEndpoint{
+			ACSEndpoint{
```
[the remark this answers is not in the document]
Comment (ebensom) [reply components/kyma-environment-broker/internal/ias/bundle.go]: 👍

File: components/kyma-environment-broker/internal/process/provisioning/ias_registration.go [hunk 7]
```diff
@@ -31,48 +29,39 @@ func (s *IASRegistrationStep) Name() string {
...
-
-	log.Info("Check if IAS ServiceProvider already exist")
-	err := spb.FetchServiceProviderData()
-	if err != nil {
-		return s.handleError(operation, err, log, "fetching IAS ServiceProvider data failed")
-	}
+	for spID := range ias.ServiceProviderInputs {
+		spb, err := s.bundleBuilder.NewBundle(operation.InstanceID, spID)
+		if err != nil {
+			return s.handleError(operation, err, log, "failed to create new ServiceProvider Bundle")
```
[the remark this answers is not in the document]
Comment (ebensom) [reply components/kyma-environment-broker/internal/process/provisioning/ias_registration.go]: ✔️

File: components/kyma-environment-broker/internal/process/deprovisioning/ias_deregistration.go [hunk 8]
```diff
@@ -29,14 +29,19 @@ func (s *IASDeregistrationStep) Name() string {
...
-	log.Info("Removing ServiceProvider from IAS")
-	err := spb.DeleteServiceProvider()
-	if err != nil {
-		msg := fmt.Sprintf("cannot delete ServiceProvider %s", spb.ServiceProviderName())
-		log.Errorf("%s: %s", msg, err)
-		return s.operationManager.RetryOperationWithoutFail(operation, msg, 5*time.Second, 5*time.Minute, log)
+	for spID := range ias.ServiceProviderInputs {
+		spb, err := s.bundleBuilder.NewBundle(operation.InstanceID, spID)
+		if err != nil {
+			return s.operationManager.OperationFailed(operation, "failed to create new ServiceProvider Bundle")
```
[the remark this answers is not in the document]
Comment (ebensom) [reply components/kyma-environment-broker/internal/process/deprovisioning/ias_deregistration.go]: ✔️

File: components/kyma-environment-broker/internal/process/provisioning/ias_type.go (hunk 1 above)
Comment (ebensom) [reply components/kyma-environment-broker/internal/process/provisioning/ias_type.go]: ✔️

File: components/kyma-environment-broker/internal/ias/bundle.go (hunk 2 above)
Comment (ebensom) [reply components/kyma-environment-broker/internal/ias/bundle.go]: Removed this, that's a remnant of a stupid idea which didn't make it's way to the PR

File: components/kyma-environment-broker/internal/ias/bundle_test.go (hunk 3 above)
Comment (ebensom) [reply components/kyma-environment-broker/internal/ias/bundle_test.go:65]: Added.

File: components/kyma-environment-broker/internal/process/deprovisioning/ias_deregistration_test.go [hunk 9]
```diff
@@ -18,14 +18,18 @@ func TestIASDeregistration_Run(t *testing.T) {
 	// given
 	memoryStorage := storage.NewMemoryStorage()
 
-	bundle := &automock.Bundle{}
-	defer bundle.AssertExpectations(t)
-	bundle.On("DeleteServiceProvider").Return(nil).Once()
-
+	bundles := map[string]*automock.Bundle{
+		"dex":     &automock.Bundle{},
+		"grafana": &automock.Bundle{},
```
[the remark this answers is not in the document]
Comment (ebensom) [reply components/kyma-environment-broker/internal/process/deprovisioning/ias_deregistration_test.go]: ✔️

File: components/kyma-environment-broker/internal/ias/input.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	ssoType       string
+	redirectPath  string
+	allowedGroups []string
+	overrides     OverridesGetter
+}
+
+var ServiceProviderInputs = map[string]ServiceProviderParam{
+	"dex": {
+		Domain:        "dex",
+		ssoType:       "saml2",
```
[the remark this answers is not in the document]
Comment (ebensom) [reply components/kyma-environment-broker/internal/ias/input.go]: Changed ✔️

File: components/kyma-environment-broker/internal/process/provisioning/ias_registration.go [hunk 11]
```diff
@@ -31,49 +31,61 @@ func (s *IASRegistrationStep) Name() string {
...
-	} else {
-		log.Infof("IAS ServiceProvider %q already registered", spb.ServiceProviderName())
-	}
 
-	log.Info("Configure IAS ServiceProvider")
-	err = spb.ConfigureServiceProvider()
-	if err != nil {
-		return s.handleError(operation, err, log, "configuring IAS ServiceProvider failed")
-	}
+		if spb.ServiceProviderType() == ias.OIDC {
```
Comment (jasiu001) [inline components/kyma-environment-broker/internal/process/provisioning/ias_registration.go:64]: Does it make sense? first if then switch?
```go
spb.ServiceProviderType() == ias.OIDC {
(...)
  switch spID {
    case ias.SPGrafanaID:
    (...)
}
```
what otherr case could be in switch?

**Comment by jasiu001** [comment]
lgtm

**Pull request merged**